src/hotspot/share/runtime/thread.cpp

```diff
@@ -2147,11 +2147,10 @@ void JavaThread::print_name_on_error(outputStream* st, char *buf, int buflen) co
 // JavaThread::print() is that we can't grab lock or allocate memory.
 void JavaThread::print_on_error(outputStream* st, char *buf, int buflen) const {
   st->print("%s \"%s\"", type_name(), get_thread_name_string(buf, buflen));
-  Thread* current = Thread::current_or_null();
-  if (current != nullptr && (!current->is_Java_thread() || JavaThread::cast(current)->is_oop_safe())) {
-    // Only access threadObj() if current thread is attached and
-    // if it is not a JavaThread or if it is a JavaThread that can safely
-    // access oops.
+  Thread* current = Thread::current();
+  if (!current->is_Java_thread() || JavaThread::cast(current)->is_oop_safe()) {
+    // Only access threadObj() if current thread is not a JavaThread
+    // or if it is a JavaThread that can safely access oops.
```

Comment:
How can a non-JavaThread safely access the oop? Is the only safe case the VMThread at a safepoint?

Reply:
The purpose of the `if (!current->is_Java_thread() ||` part of the if-statement
is to allow the code to work as it did before for the non-JavaThread case. Before
this fix, if a non-JavaThread called into this code, then it was allowed to execute
this code. I've preserved that behavior and I've see no failures that indicate that
this is a problem.

Do I know what non-JavaThreads might wander in here? No I don't.

Reply:
I realise this is pre-existing behaviour. It seems odd that a non-JavaThread can touch an oop when an exiting JavaThread cannot - how are they different? Or do we just cross our fingers and hope for the best with a non-JavaThread because it is rare? Perhaps @fisk can explain?